Update/update to multi license assignment

oikeme opened the pull request:
#### Proposed Changes

* This PR builds on updates to [support multi license assignment](https://github.com/Automattic/wp-calypso/pull/69864) and updates the license assignment page to support this in the assignment process

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Apply this PR
* Run yarn-start-jetpack-cloud
* Visit the licenses page at `http://jetpack.cloud.localhost:3000/partner-portal/licenses` and attempt to assign a single license. Ensure that this works as expected and the banner indicating license assignment is displayed.
* Visit the page again and attempt to assign a license but first copy a second unassigned license from the page first 
* Use the Assign License button on another license aside from the copied license
* Once on the license assignment page update the `key` query param to `products`
* Use the Assign 2 licenses button to assign the licenses
* Ensure both licenses are assigned and the banner is displayed indicating both licenses have been assigned. 

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

oikeme committed:
Update the license assignment page

Update the license assignment page query param for multi license from "keys" to "products" Also update all instances of singular license to plural when the products query param is used and there is more than one product license being issued

oikeme committed:
Update license assignment

Update license assignment to support assigning multiple licenses to a site

oikeme committed:
Merge branch 'trunk' into update/update-to-multi-license-assignment

oikeme committed:
Split assign multiple licenses mutation

Create a new mutation to allow for assigning multiple licenses to a site

oikeme committed:
Merge branch 'trunk' into update/update-to-multi-license-assignment

yashwin commented on client/jetpack-cloud/sections/partner-portal/hooks.tsx:
```diff
@@ -514,3 +514,73 @@ export function useIssueMultipleLicenses(
+	const isLoading = assignLicense.isLoading;
+	const selectedSiteId = selectedSite?.ID as number;
+	const assign = useCallback( async () => {
+		// setIsSubmitting( true );
```
I feel like the whole code written in this function can be moved into a different reusable function and used here and in `useIssueMultipleLicenses` since we are doing the same thing there. Let me know your thoughts.

yashwin commented on client/jetpack-cloud/sections/partner-portal/hooks.tsx:
```diff
@@ -514,3 +514,73 @@ export function useIssueMultipleLicenses(
+	} );
+	const dispatch = useDispatch();
+	const assignLicense = useAssignLicenseMutation( {
+		onSuccess: () => {
```
`onSuccess` & `onError` can be removed once we create the reusable assign function since it is handled there. We also need to redirect the user to the dashboard if the source is dashboard

yashwin commented on client/jetpack-cloud/sections/partner-portal/assign-license-step-progress/index.tsx:
```diff
@@ -48,14 +48,14 @@ export default function AssignLicenseStepProgress( { currentStep }: Props ) {
 	const sites = useSelector( getSites ).length;
 
-	const steps: Step[] = [ { key: 'issueLicense', label: translate( 'Issue new license' ) } ];
+	const steps: Step[] = [ { key: 'issueLicense', label: translate( 'Issue new licenses' ) } ];
```
I guess this change is copied from the previous base branch, which we decided to add later since multiple license flow is not enabled in production.

yashwin commented on client/jetpack-cloud/sections/partner-portal/assign-license-step-progress/index.tsx:
```diff
@@ -48,14 +48,14 @@ export default function AssignLicenseStepProgress( { currentStep }: Props ) {
 
 	if ( sites > 0 ) {
-		steps.push( { key: 'assignLicense', label: translate( 'Assign license' ) } );
+		steps.push( { key: 'assignLicense', label: translate( 'Assign licenses' ) } );
```
I guess this change is copied from the previous base branch, which we decided to add later since multiple license flow is not enabled in production.

yashwin commented on client/jetpack-cloud/sections/partner-portal/assign-license-form/index.tsx:
```diff
@@ -46,14 +43,15 @@ export default function AssignLicenseForm( {
 	const licenseKeysArray = products !== undefined ? products.split( ',' ) : [ licenseKey ];
 	const onSelectSite = ( site: any ) => setSelectedSite( site );
-	const notificationId = 'partner-portal-assign-license-form';
+	const [ onAssignLicense, isLoading ] = useAssignMultipleLicenses(
```
How about we change `onAssignLicense` to `assignLicenses`?

yashwin reviewed (changes requested):
This works as described in the testing instructions, but I have added some comments.

While testing, I found one issue: the radio button on the assign license page is not working properly. Even though I have selected a site, it doesn't highlight the selected one. 

<img width="1084" alt="Screenshot 2022-11-15 at 12 37 56 PM" src="https://user-images.githubusercontent.com/10586875/201852910-5961f34e-6fed-42eb-b3bc-e0062cca7be6.png">

oikeme committed:
Merge branch 'trunk' into update/update-to-multi-license-assignment

oikeme committed:
Update assign licenses function

Update the assign licenses mutation to use a separate function that can later be used when issuing multiple licenses. Also fix the radio box no longer being selected when assigning a license

oikeme committed:
Merge branch 'trunk' into update/update-to-multi-license-assignment

yashwin reviewed (commented):
I tested the changes again, and it works as expected. But I see a couple of errors.

<img width="1011" alt="Screenshot 2022-11-16 at 2 57 19 PM" src="https://user-images.githubusercontent.com/10586875/202142227-b50b04e7-c684-4f7f-b451-898bd418512c.png">
<img width="956" alt="Screenshot 2022-11-16 at 2 57 28 PM" src="https://user-images.githubusercontent.com/10586875/202142253-f7e4120e-f179-4fd6-ab04-52052c9c3efc.png">

yashwin commented on client/jetpack-cloud/sections/partner-portal/hooks.tsx:
```diff
@@ -280,6 +280,80 @@ export function useLicenseIssuing(
+	} );
+	const isLoading = assignLicense.isLoading;
+	const selectedSiteId = selectedSite?.ID as number;
+	const assignMultipleLIcenses = useCallback( async () => {
```
There is a typo here, it should be `assignMultipleLicenses`

yashwin commented on client/jetpack-cloud/sections/partner-portal/hooks.tsx:
```diff
@@ -280,6 +280,80 @@ export function useLicenseIssuing(
+		licenseKeys.forEach( ( licenseKey ) => {
+			dispatch(
+				recordTracksEvent( 'calypso_partner_portal_assign_license_submit', {
+					license_key: licenseKey,
```
How about we do an event tracking for all the `licenseKeys` at once rather than doing it for individual licenses?

```
dispatch(
			recordTracksEvent( 'calypso_partner_portal_assign_multiple_licenses_submit', {
				products: licenseKeys.join( ',' ),
			} )
);
```

oikeme committed:
Update type hinting and fix typo

Update type hinting for assign licenses function and fix a typo with function name. Also updated the tracks event record to batch record multi license assignment

oikeme committed:
Merge branch 'trunk' into update/update-to-multi-license-assignment

yashwin commented on client/jetpack-cloud/sections/partner-portal/assign-license-form/index.tsx:
```diff
@@ -46,14 +43,16 @@ export default function AssignLicenseForm( {
-	const notificationId = 'partner-portal-assign-license-form';
+	const onSelectSite = ( site: any ) => {
+		setSelectedSite( site );
+		setSelectedSiteId( site?.ID );
```
Do we need a separate state for ID? How about we do selectedSite.ID?

yashwin reviewed (approved):
LGTM. It works as mentioned in the testing instructions. I have left a small comment.

oikeme committed:
Update index.tsx

Remove redundant constant for siteId

oikeme committed:
Merge branch 'trunk' into update/update-to-multi-license-assignment

oikeme committed:
Merge branch 'trunk' into update/update-to-multi-license-assignment
